assets/js/base/context/cart-checkout/checkout-state/index.js

```diff
@@ -0,0 +1,203 @@
+/**
+ * External dependencies
+ */
+import {
+	createContext,
+	useContext,
+	useReducer,
+	useRef,
+	useMemo,
+	useEffect,
+} from '@wordpress/element';
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import { actions } from './actions';
+import { reducer } from './reducer';
+import { DEFAULT_STATE, STATUS } from './constants';
+import {
+	EMIT_TYPES,
+	emitterSubscribers,
+	emitEvent,
+	emitEventWithAbort,
+	reducer as emitReducer,
+} from './event-emit';
+
+/**
+ * @typedef {import('@woocommerce/type-defs/checkout').CheckoutDispatchActions} CheckoutDispatchActions
+ * @typedef {import('@woocommerce/type-defs/contexts').CheckoutDataContext} CheckoutDataContext
+ */
+
+const CheckoutContext = createContext( {
+	submitLabel: '',
+	onSubmit: () => void null,
+	isComplete: false,
+	isIdle: false,
+	isCalculating: false,
+	isProcessing: false,
+	hasError: false,
+	redirectUrl: '',
+	orderId: 0,
+	onCheckoutCompleteSuccess: ( callback ) => void callback,
+	onCheckoutCompleteError: ( callback ) => void callback,
+	onCheckoutProcessing: ( callback ) => void callback,
+	dispatchActions: {
+		resetCheckout: () => void null,
+		setRedirectUrl: ( url ) => void url,
+		setHasError: ( hasError ) => void hasError,
+		incrementCalculating: () => void null,
+		decrementCalculating: () => void null,
+		setOrderId: ( id ) => void id,
+	},
+	hasOrder: false,
+	isCart: false,
+} );
+
+/**
+ * @return {CheckoutDataContext} Returns the checkout data context value
+ */
+export const useCheckoutContext = () => {
+	return useContext( CheckoutContext );
+};
+
+/**
+ * Checkout state provider
+ * This provides provides an api interface exposing checkout state for use with
+ * cart or checkout blocks.
+ *
+ * @param {Object}  props                     Incoming props for the provider.
+ * @param {Object}  props.children            The children being wrapped.
+ * @param {string}  props.redirectUrl         Initialize what the checkout will
+ *                                            redirect to after successful
+ *                                            submit.
+ * @param {string}  props.submitLabel         What will be used for the checkout
+ *                                            submit button label.
+ * @param {boolean} props.isCart              If context provider is being used
+ *                                            in cart context.
+ */
+export const CheckoutStateProvider = ( {
```

Comment:
We could maybe call this CheckoutOrderProvider? Most of the interfaces are about preparing data for an order or the order itself  .. not sure if this is a better name.

Reply:
Hmm, ya there _is_ order data exposed on it (if available) but that's not the primary purpose of this provider.